### DISCUSSION TITLE
[Userobjects]

### DISCUSSION THREAD

**Discussion by coskrrb2002** [idaholab/moose#21497] [Q&A General] [answered]
Dear users,

I am trying to calculate a value by using Userobjects.

Initial guessing of value A -> Use A to calculate quadrature point values -> Sum up all the values on a plane (or surface) -> If this value is not meet with some criteria, do it again by changing value A -> ... ->Until satisfy the criteria.

Are there any Userobjects to support this method? or any other ways?

Best regards,

**Comment by GiudGiud** [comment, marked as answer]
Hello

How often you trying to perform these iterations? Is this one iteration per time step or are you trying to converge the process every time step (or more often, every nonlinear/linear iteration) ?

Is A a field or a single floating point value?

Guillaume

**Reply by coskrrb2002** [reply]
Dear Guillaume,


I just have to perform this calculation once per time step at timestep_begin.

A should be a single floating value, it doesn't have to be a field value.


Best regards,
Nakkyu

**Reply by GiudGiud** [reply]
I think the easiest setup will be to place this A calculation in a MultiApp.

This multiapp will run a fake transient to determine A.
You will want to use this
https://mooseframework.inl.gov/source/postprocessors/SideAverageValue.html
to compute the value on the plane (add a sideset for the plane). 
then you will want to use the control system to modify the value of A at each iteration:
this is the base one you could use
https://mooseframework.inl.gov/source/controls/RealFunctionControl.html
but I tend to think you may want to use this one instead, which I wrote
https://mooseframework.inl.gov/source/controls/PIDTransientControl.html

**Reply by coskrrb2002** [reply]
Thank you very much Guillaume.

 I will give it a try!

**Reply by coskrrb2002** [reply]
Is it possible to do this job by only using a SideIntegralPostprocessor?

I know the SideIntegralPostprocessor supports the summation of values along with surfaces.

But, it seems like SideIntegralPostprocessor cannot do the calculation multipletimes like the below:

Initial A value -> Calculate B on each quadrature point and sum all B values into value C (C = summation of B[_qp] for all surfaces) based on A value -> If B value is not met criteria, then update A -> Repeat until B is fulfilled the criteria

I tried to make it as below, but I figured out this doesn't sum up all surfaces. This code just adds one element. (I guess).

Because I can't understand how to apply threadJoin().

Can you give me some advice about it?
```
-------------------------------------------------------------------------------------------------------------------
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#include "EcorrPost.h"

#include "libmesh/quadrature.h"

//registerMooseObject("corrosionApp", EcorrPost);

InputParameters
EcorrPost::validParams()
{
  InputParameters params = SidePostprocessor::validParams();
  params.addCoupledVar("T",298.15,"T");
  params.addRequiredCoupledVar("C9","HS-");
  params.addCoupledVar("C1",0,"C1");
  params.addCoupledVar("C0",0,"C0");
  params.addCoupledVar("C3",0,"C3");
  params.addCoupledVar("C6",0,"C6");
  return params;
}

EcorrPost::EcorrPost(const InputParameters & parameters)
  : SidePostprocessor(parameters),
	_qp(0),
	_integral_value(0),
	_qp_integration(true),
	_integral_value2(0),
	_OP(0),
	_Isum(0),
    _T(coupledValue("T")),
    _C9(coupledValue("C9")),
    _C1(coupledValue("C1")),
    _C0(coupledValue("C0")),
    _C3(coupledValue("C3")),
    _C6(coupledValue("C6")),
    _mesh(_subproblem.mesh().getMesh())
{
}

void
EcorrPost::initialize()
{
  _integral_value = 0;
  _integral_value2 = 0;
  _Isum = 0;
}

void
EcorrPost::execute()
{
	_Isum += computeCurrents();
}

Real
EcorrPost::getValue()
{
  _OP = -0.2;
  std::cout << "Initial Isum" << _Isum << std::endl;
  gatherSum(_Isum);

  Real aaa = _current_side;
  std::cout << "Current node id:" << aaa << std::endl;
  std::cout << "Gather Isum" << _Isum << std::endl;
  Real bbb = _mesh.n_active_elem();
  Real ccc = _mesh.n_elem();
  std::cout << "Active elements number:" << bbb << "Total elements number" << ccc << std::endl;

  for (int i = 0; i < 5; i++)
  {
  if (_Isum > 1E-8)
  {
	  initialize();
	  _OP = _OP - 1E-5;
	  execute();
	  gatherSum(_Isum);
	  std::cout << "Isum" << _Isum << std::endl;
  }
  else if (_Isum < -1E-8)
  {
	  initialize();
	  _OP = _OP + 1E-5;
	  execute();
	  gatherSum(_Isum);
	  std::cout << "Isum" << _Isum << std::endl;
  }
  else
	  break;

  }

  return _Isum;
}

void
EcorrPost::threadJoin(const UserObject & y) //As in the MOOSE index pages, it should not be used! Just use gathersum
{
  const EcorrPost & pps = static_cast<const EcorrPost &>(y);
  _integral_value += pps._integral_value;
  _integral_value2 += pps._integral_value2;
  _Isum += pps._Isum;
}

Real
EcorrPost::computeIntegral()
{
	return 0;
}

Real
EcorrPost::computeCurrents()
{
  Real sum = 0;
    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
    {
      sum = _C1[_qp] + _C0[_qp] + _C3[_qp] + _C6[_qp] + _C9[_qp] + sum;
    }
  return sum;
}
```
-------------------------------------------------------------------------------------------------------------------

**Reply by GiudGiud** [reply]
no this isnt going to work. 
The postprocessors are part of a loop over all elements, which is in the internals of moose (the loops are in include/loops). 
What you did is not executing the postprocessor over the entire mesh several times, only over a single element

You need to decompose operations, and use a different object for each part of the process

**Reply by coskrrb2002** [reply]
Thank you for your comment!

As following your comments, multiapp + postprocessor works.

Now I am trying to make an executioner file that can terminate the calculation of value A!

Thank you very much!

**Reply by GiudGiud** [reply]
ok so to do that you may use a Terminator and store the A value as a postprocessor
https://mooseframework.inl.gov/source/userobject/Terminator.html

**Reply by coskrrb2002** [reply]
Yes, thanks a lot! I found a terminator and modify it to finish the calculation between -1 < A < 1, and it works very well!

Cheers,
nakkyu
